File: lib/nest/messages/think_tags.ex
defmodule Nest.Messages.ThinkTags do
  @moduledoc """
  Strip `<think>...</think>` content from text. Used by the
  compactor's LLM call (see `Nest.Agents.Agent.Compaction`):
  the compactor sends the agent's prior conversation to the
  summarization LLM and wants to remove the prior turn's
  reasoning content so the compactor sees clean text (the
  reasoning already happened — repeating it inflates the
  summary without adding signal). The summary itself is
  stored intact; only the LLM call's input is stripped.

  ## Behavior

  `<think>...</think>` blocks are removed from the text. The
  markers themselves are stripped; the inner content goes to
  `:none`. Surrounding text is preserved.

    * No markers → returns the input unchanged.
    * One block → returns the text with the block removed.
    * Multiple blocks → returns the text with each block removed.
    * Nested `<think>` inside a think block → the inner
      `<think>` is treated as text inside the think block
      (no recursion), but its matching `</think>` is
      consumed to keep the outer block from closing early.
      The outer block closes only at depth 0.
    * Empty think block (`<think></think>`) → removed (zero
      contribution).
    * Orphan `</think>` (no matching `<think>`) → the orphan
      and everything from the orphan to the end of the string
      is removed. Stray `</think>` text in a response is
      model reasoning, not user-visible content — it should
      not reach the compactor LLM.
    * Orphan `<think>` (no matching `</think>`) → everything
      from the orphan to the end of the string is removed
      (an incomplete think block — also reasoning content).

  ## Relationship to JS

  Mirrors `assets/js/utils/thinkTags.js`'s `splitThinkTags`:
  the JS side splits so the UI can render the thinking as a
  collapsed block; the Elixir side strips so the LLM sees
  clean text. Same algorithm, opposite direction.
  """

  @opening_tag "<think>"
  @closing_tag "</think>"
  @opening_tag_len byte_size(@opening_tag)
  @closing_tag_len byte_size(@closing_tag)

  @doc """
  Strip `<think>...</think>` blocks from `text`. See the
  module doc for the full behavior.
  """
  @spec strip(String.t()) :: String.t()
  def strip(text) when is_binary(text) do
    strip(text, 0, [])
  end

  def strip(_), do: ""

  # `depth` is the number of `<think>` openers minus `</think>`
  # closers we've seen. depth == 0 means we're in text mode;
  # depth > 0 means we're inside one or more think blocks.
  # `acc` is the reverse-order IO list of completed text runs.
  defp strip("", _depth, acc), do: finalize(acc)

  defp strip(text, 0, acc) do
    case next_in_text_mode(text) do
      :none ->
        finalize(acc) <> text

      {:open, idx} ->
        before_marker = binary_part(text, 0, idx)
        new_acc = push_if_nonempty(acc, before_marker)
        tail = tail_after(text, idx, @opening_tag_len)
        strip(tail, 1, new_acc)

      {:orphan_close, idx} ->
        before_marker = binary_part(text, 0, idx)
        new_acc = push_if_nonempty(acc, before_marker)
        finalize(new_acc)
    end
  end

  defp strip(text, depth, acc) when depth > 0 do
    case next_in_thinking_mode(text) do
      :none ->
        # No more markers — discard the rest (it's all reasoning).
        finalize(acc)

      {:close, idx} ->
        tail = tail_after(text, idx, @closing_tag_len)
        strip(tail, depth - 1, acc)

      {:open, idx} ->
        tail = tail_after(text, idx, @opening_tag_len)
        strip(tail, depth + 1, acc)
    end
  end

  # Find the next marker in text mode (depth = 0). Returns
  # `:none` if no markers; otherwise `{:open, idx}` for a
  # `<think>` opener or `{:orphan_close, idx}` for an orphan
  # `</think>` (treated as the start of a thinking segment).
  defp next_in_text_mode(text) do
    open_idx = match_index(text, @opening_tag)
    close_idx = match_index(text, @closing_tag)

    case {open_idx, close_idx} do
      {nil, nil} -> :none
      {oi, nil} -> {:open, oi}
      {nil, ci} -> {:orphan_close, ci}
      {oi, ci} when oi <= ci -> {:open, oi}
      {_oi, ci} -> {:orphan_close, ci}
    end
  end

  # Find the next marker in thinking mode (depth > 0). Returns
  # `:none` if no markers; otherwise the first marker by
  # index. `<think>` opens a nested level (depth+1), `</think>`
  # closes the current level (depth-1). Whichever comes first
  # wins — depth tracking ensures the matching `</think>` is
  # consumed before a stray `</think>` later in the text can
  # close the outer block.
  defp next_in_thinking_mode(text) do
    open_idx = match_index(text, @opening_tag)
    close_idx = match_index(text, @closing_tag)

    case {open_idx, close_idx} do
      {nil, nil} -> :none
      {oi, nil} -> {:open, oi}
      {nil, ci} -> {:close, ci}
      {oi, ci} when oi <= ci -> {:open, oi}
      {_oi, ci} -> {:close, ci}
    end
  end

  defp finalize(acc), do: acc |> :lists.reverse() |> IO.iodata_to_binary()

  defp push_if_nonempty(acc, ""), do: acc
  defp push_if_nonempty(acc, str), do: [str | acc]

  defp tail_after(text, idx, marker_len) do
    tail_start = idx + marker_len
    size = byte_size(text)
    binary_part(text, tail_start, size - tail_start)
  end

  defp match_index(text, marker) do
    case :binary.match(text, marker) do
      {idx, _} -> idx
      :nomatch -> nil
    end
  end

  @doc """
  Walk a message's `parts` list, rewriting each `Part.Text`
  to strip `<think>...</think>` content from its `text`
  field. All other part kinds pass through unchanged. Returns
  the rewritten parts list (same length as input).
  """
  @spec strip_from_parts([Nest.Messages.Part.t()]) :: [Nest.Messages.Part.t()]
  def strip_from_parts(parts) when is_list(parts) do
    Enum.map(parts, fn
      %Nest.Messages.Part.Text{text: text} = part ->
        %{part | text: strip(text)}

      part ->
        part
    end)
  end

  def strip_from_parts(_), do: []
end

File: assets/js/utils/thinkTags.js
/**
 * Splits a string on `<think>` / `</think>` markers and
 * returns the segments in original order. Each segment is
 * `{kind: "text" | "thinking", text: string}`.
 *
 * `<think>...</think>` blocks become `{kind: "thinking", text}`.
 * Everything else is `{kind: "text", text}`.
 *
 * Behavior:
 *
 *   - No markers: returns a single text segment.
 *   - One block: returns `[text-before, thinking, text-after]`.
 *   - Multiple blocks: returns the interleaved sequence.
 *   - Nested `<think>` inside a think block: the inner
 *     `<think>` is treated as text inside the think block
 *     (no recursion into a new thinking segment), but its
 *     matching `</think>` is consumed to keep the outer
 *     block from closing early.
 *   - Empty think blocks (`<think></think>`): emitted as a
 *     thinking segment with `text: ""` and then collapsed to
 *     "no thinking" downstream if needed.
 *   - Orphan `</think>` (no matching `<think>`): the orphan
 *     is treated as a thinking marker — everything from the
 *     orphan to the end of the string (or the next
 *     `<think>`) is routed to the thinking channel. This
 *     matches the "with any `<think>` blocks intact"
 *     principle: stray `</think>` text in the response is
 *     model reasoning, not user-visible content.
 *   - Orphan `<think>` (no matching `</think>`): everything
 *     from the orphan to the end of the string is routed to
 *     the thinking channel (incomplete think block).
 */
export function splitThinkTags(text) {
  if (typeof text !== "string" || text.length === 0) return [];
  return splitThinkTagsInner(text);
}

const OPEN = "<think>";
const CLOSE = "</think>";
const OPEN_LEN = OPEN.length;
const CLOSE_LEN = CLOSE.length;

function splitThinkTagsInner(text) {
  const segments = [];
  let buffer = "";
  let inThinking = false;
  let depth = 0;
  let i = 0;

  const flush = () => {
    // Always emit, even an empty segment — an empty think
    // block (`<think></think>`) is a real thinking segment
    // and gets dropped downstream by the caller. Empty text
    // segments (zero-length runs of text) are dropped here.
    if (buffer.length === 0 && !inThinking) return;
    segments.push({ kind: inThinking ? "thinking" : "text", text: buffer });
    buffer = "";
  };

  while (i < text.length) {
    if (inThinking) {
      // Walk through the text, tracking nested depth.
      let openIdx = text.indexOf(OPEN, i);
      let closeIdx = text.indexOf(CLOSE, i);
      while (i < text.length) {
        if (openIdx === -1 && closeIdx === -1) {
          buffer += text.slice(i);
          i = text.length;
          break;
        }
        if (openIdx !== -1 && (closeIdx === -1 || openIdx <= closeIdx)) {
          buffer += text.slice(i, openIdx + OPEN_LEN);
          depth += 1;
          i = openIdx + OPEN_LEN;
          openIdx = text.indexOf(OPEN, i);
        } else {
          // closeIdx comes first
          if (depth === 1) {
            buffer += text.slice(i, closeIdx);
            flush();
            i = closeIdx + CLOSE_LEN;
            inThinking = false;
            depth = 0;
            break;
          }
          buffer += text.slice(i, closeIdx + CLOSE_LEN);
          depth -= 1;
          i = closeIdx + CLOSE_LEN;
          closeIdx = text.indexOf(CLOSE, i);
        }
      }
      continue;
    }

    // text channel: find the next opening `<think>` or an
    // orphan `</think>`. Whichever comes first wins.
    const openIdx = text.indexOf(OPEN, i);
    const orphanCloseIdx = text.indexOf(CLOSE, i);
    let nextIdx = -1;
    let nextKind = null;

    if (openIdx !== -1 && orphanCloseIdx !== -1) {
      if (openIdx <= orphanCloseIdx) {
        nextIdx = openIdx;
        nextKind = "open";
      } else {
        nextIdx = orphanCloseIdx;
        nextKind = "orphan_close";
      }
    } else if (openIdx !== -1) {
      nextIdx = openIdx;
      nextKind = "open";
    } else if (orphanCloseIdx !== -1) {
      nextIdx = orphanCloseIdx;
      nextKind = "orphan_close";
    }

    if (nextIdx === -1) {
      buffer += text.slice(i);
      i = text.length;
    } else {
      buffer += text.slice(i, nextIdx);
      flush();
      if (nextKind === "open") {
        i = nextIdx + OPEN_LEN;
        inThinking = true;
        depth = 1;
      } else {
        i = nextIdx + CLOSE_LEN;
        inThinking = true;
        depth = 1;
      }
    }
  }

  flush();
  return segments;
}

/**
 * Walks a message's `parts` array. For each `Part.Text`, splits
 * the text on `<think>` / `</think>` markers, collects the
 * thinking content into a single string, and returns the
 * remaining text segments as separate `Part.Text` entries
 * (preserving the original `text` field for any non-`<think>`
 * content). `Part.Thinking` passes through — its content
 * accumulates into the same `thinking` string. Other part
 * types pass through unchanged in their original position.
 *
 * Returns:
 *   {
 *     thinking: string | null,    // concatenated thinking, or null if empty
 *     textParts: Part[],          // parts list with <think> blocks split out
 *   }
 *
 * Used by the history pane and the active message list to
 * render `` content as a collapsed ThinkingBlock above
 * the visible reply.
 */
export function splitThinkFromParts(parts) {
  if (!Array.isArray(parts)) return { thinking: null, textParts: [] };

  const textParts = [];
  let thinking = null;

  for (const p of parts) {
    if (!p) continue;

    if (p.kind === "thinking") {
      const t = p.thinking || "";
      if (t) {
        thinking = (thinking || "") + (thinking ? "\n\n" : "") + t;
      }
      continue;
    }

    if (p.kind === "text") {
      const segments = splitThinkTags(p.text || "");
      for (const seg of segments) {
        if (seg.kind === "thinking") {
          thinking = (thinking || "") + (thinking ? "\n\n" : "") + seg.text;
        } else {
          textParts.push({ ...p, text: seg.text });
        }
      }
      continue;
    }

    textParts.push(p);
  }

  return { thinking, textParts };
}
